feat(stream): add Stream server & supervisor

OCannings opened the pull request:
The `Juicebox.Stream.Server` and `Juicebox.Stream.Supervisor` modules
have been introduced.

The Server API is as follows:

_.add(stream_id, video)_
Plays the video if one is not already playing, otherwise adds it to the
queue.

_.skip(stream_id)_
Skips the currently playing video

_.remaining_time(stream_id)_
Returns (in ms) the playback time remaining for the current video

_.playing(stream_id)_
Returns the currently playing video (%Juicebox.Youtube.Video{})

Videos will continue to automatically play until the queue is empty.
Adding additional videos will auto-restart playback.

TODO:
- Connect Stream to Phoenix app
- Support ordered stream (by upvotes)

OCannings committed:
feat(stream): add Stream server & supervisor

The `Juicebox.Stream.Server` and `Juicebox.Stream.Supervisor` modules
have been introduced.

The Server API is as follows:

*.add(stream_id, video)*
Plays the video if one is not already playing, otherwise adds it to the
queue.

*.skip(stream_id)*
Skips the currently playing video

*.remaining_time(stream_id)*
Returns (in ms) the playback time remaining for the current video

*.playing(stream_id)*
Returns the currently playing video (%Juicebox.Youtube.Video{})

Videos will continue to automatically play until the queue is empty.
Adding additional videos will auto-restart playback.

TODO:
* Connect Stream to Phoenix app
* Support ordered stream (by upvotes)

ljkbennett commented on lib/juicebox/stream.ex:
```diff
@@ -0,0 +1,37 @@
+    {:ok, []}
+  end
+
+  def get_queue_length do
```
Is this a common pattern? i.e. to add helper methods to the module which then just do GenServer calls? I think if I was coming at this fresh I'd have 2 modules, one to be the GenServer and one to be the one that calls or casts to the gen server.

OCannings committed:
docs(Stream): add basic documentation

OCannings committed:
feat(Stream): add Stream.Server.queue/1

`.queue(stream_id)` now returns a list of the videos in the queue

OCannings committed:
feat(Stream): add Stream.vote/2

The Stream server now creates a queue of `%Juicebox.Stream.Track{}`s
rather than `%Juicebox.Youtube.Video{}`s. This allows us to attach
non-video related attributes.

Calling `Stream.vote/2' with a stream id and track ID will increment the
number of votes for the given track.

The queue is now ordered by number of votes.

ljkbennett commented:
🍙
